Docs: Versioning long running functions (and determinism) explanation (#790)
inngest/website: pages/docs/learn/versioning.mdx

export const description = "Learn how ."

# Versioning long running functions

> It's inevitable that long-running functions change over time. Inngest allows developers to implement multiple strategies for changing long-running code over time. In order to understand how to best manage changes over time, it's important to understand how the SDK implements determinism and [executes steps](/docs/learn/how-functions-are-executed).

Reviewer: ```suggestion
Long-running functions inevitably change over time. Inngest enables developers to implement multiple strategies for changing long-running code over time. To manage these changes effectively, it's crucial to understand how the SDK implements determinism and [executes steps](https://chatgpt.com/docs/learn/how-functions-are-executed).
```

- changes the first sentence to be clearer and more impactful
- solves redundant repetition in the last sentence

Author: ```suggestion
ong-running functions inevitably change over time. Inngest enables developers to implement multiple strategies for changing long-running code over time. To manage these changes effectively, it's crucial to understand how the SDK implements determinism and [executes steps](/docs/learn/how-functions-are-executed).
```

## Determinism in functions

> Determinism is the same in *every* Inngest language SDK. Except for language-specific idioms, all SDKs implement the same logic. In all SDKs, functions in Inngest are a series of steps. Each step runs reliably, will retry on failure, and is as close to exactly once as possible (outbound network failure when reporting completing steps notwithstanding).

Reviewer: ```suggestion
Determinism is consistent in *every* Inngest language SDK. Except for language-specific idioms, all SDKs implement the same logic. In every SDK, functions in Inngest are a series of steps. Each step runs reliably, will retry on failure, and is as close to exactly-once execution as possible (excluding outbound network failure when reporting completing steps).
```

### How the SDK works with steps

As covered in [_How Inngest functions are executed_](/docs/learn/how-functions-are-executed), each step in a function has a unique identifier, represented as a string. Each time a step is found, the SDK checks whether the step has been executed. It does this by:

> 1. Hashing the step's unique identifier along with an counter of the number of times the step has been called. This allows steps to be used in a loop.

Reviewer: ```suggestion
1. Hashing the step's unique identifier along with a counter of the number of times the step has been called. In this way, steps can be used in a loop.
```

Author: ```suggestion
1. Hashing the step's unique identifier along with a counter of the number of times the step has been called. This enables steps to be used in a loop.
```
Riff on the edit for clarity of what this does

    2. If the hash isn't found, the SDK executes the step and returns the output to Inngest to be stored in the function run state.

After a step completes, the function execution immediately ends. The function is re-executed from the top with the updated memoized state until the function completes.

### Handling determinism

The SDK handles determinism *gracefully* by default. The SDK keeps track of the order in which every step is executed. If new steps are added, they're executed when they're first discovered. This means that:

- The SDK always knows if functions are deterministic, even over months or years.
- **New steps, or steps with changed IDs, are executed when they're discovered.** If the order of step executions change, a warning is logged by default{/* (your functions can be made to permanently fail by enabling strict mode)*/}. Logging a warning allows you to comfortably extend and improve functions over time, without worrying about in-progress functions failing completely or panicking.

## Change management across versions

Given the above, there are a few strategies for change management:

> - **Adding new steps to a function is generally safe.** New steps will be executed when the functions re-run (after a step completes). If a function has step `[A, B, C]` and you add a new step `Z` in-between the first steps, the executor will run steps `[A, Z, B, C]` and log a warning. The caveat here is that you must take care to ensure that the new step can run out-of-order and doesn't reference undefined variables. Note that step `B` and `C` will *not* automatically re-run. Instead, a warning will be logged by default. You can change logging a warning and instead permanently fail by enabling strict mode. Permanently failing runs is acceptable, and you can use replay to bulk-replay permanent failures.

Reviewer: ```suggestion
- **Adding new steps to a function is generally safe.** New steps will be executed when the functions re-run (after a step completes). Imagine a function has steps `[A, B, C]`. When you add a new step `Z` in-between the first steps, the executor will run steps `[A, Z, B, C]` and log a warning. The caveat here is that you must take care to ensure that the new step can run out-of-order and doesn't reference undefined variables. Note that step `B` and `C` will *not* automatically re-run. Instead, a warning will be logged by default. You can change logging a warning and instead permanently fail by enabling strict mode. Failing runs permanently is acceptable, and you can use [Replay](/docs/platform/replay) to bulk-replay permanent failures.
```

> - **Forcing steps to re-run by changing step IDs.** This changes the hash, which forces re-evaluation as the step's state is not found. Note that the SDK will log a warning by default as the order of step execution changes. If you change step `C`'s ID to `Z`, your run's state will expect steps `[A, B, C]` to run and instead will see `[A, B, Z]`.

Reviewer: ```suggestion
- **Forcing steps to re-run by changing step IDs.** This changes the hash, which forces re-evaluation as the step's state is not found. Note that the SDK will log a warning by default as the order of step execution changes. If you change step `C`'s ID to `F`, your run's state will expect steps `[A, B, C]` to run and instead will see `[A, B, F]`.
```

changing so folks know this is not a continuation of the previous point
